File: Tests/SourceStateTests.cs
using System;
using Stringier.Patterns;
using static Stringier.Patterns.Pattern;
using Xunit;

namespace Tests {
	public class SourceStateTests {
		[Fact]
		public void StoreAndRestore() {
			Source source = new Source("hello world");
			SourceState state = source.Store();
			ResultAssert.Captures("hello", "hello".Consume(ref source));
			source.Restore(state);
			ResultAssert.Captures("hello", "hello".Consume(ref source));
		}
	}
}

File: Tests.FSharp/SourceStateTests.fs
namespace Patterns

open System
open Stringier.Patterns
open Defender
open Xunit

type SourceStateTests() =
    inherit Trial()

    [<Fact>]
    let ``store and restore`` () =
        let mutable source = Source("hello world")
        let state = source.Store()
        let mutable result = "hello".Consume(&source)
        Assert.True(result.Success)
        source.Restore(state)
        //The only way this can succeed is if the restore worked
        Assert.True(result.Success)